// create-nx-workspace/src/utils/error-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapErrorToBodyLines = exports.CreateNxWorkspaceError = void 0;
class CreateNxWorkspaceError extends Error {
    constructor(logMessage, code, logFile) {
        super(logMessage);
        this.logMessage = logMessage;
        this.code = code;
        this.logFile = logFile;
        this.name = 'CreateNxWorkspaceError';
    }
}
exports.CreateNxWorkspaceError = CreateNxWorkspaceError;
function mapErrorToBodyLines(error) {
    const errorLines = error.message?.split('\n').filter((line) => !!line.trim());
    if (errorLines.length < 3) {
        const lines = [`Error: ${error.message}`];
        if (process.env.NX_VERBOSE_LOGGING) {
            lines.push(`Stack: ${error.stack}`);
        }
        return lines;
    }
    const lines = error instanceof CreateNxWorkspaceError
        ? [`Exit code: ${error.code}`, `Log file: ${error.logFile}`]
        : [];
    if (process.env.NX_VERBOSE_LOGGING) {
        lines.push(`Error: ${error.message}`);
        lines.push(`Stack: ${error.stack}`);
    }
    return lines;
}
exports.mapErrorToBodyLines = mapErrorToBodyLines;
