// create-nx-workspace/src/utils/preset/preset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isKnownPreset = exports.Preset = void 0;
var Preset;
(function (Preset) {
    Preset["Apps"] = "apps";
    Preset["NPM"] = "npm";
    Preset["TS"] = "ts";
    Preset["WebComponents"] = "web-components";
    Preset["AngularMonorepo"] = "angular-monorepo";
    Preset["AngularStandalone"] = "angular-standalone";
    Preset["ReactMonorepo"] = "react-monorepo";
    Preset["ReactStandalone"] = "react-standalone";
    Preset["VueMonorepo"] = "vue-monorepo";
    Preset["VueStandalone"] = "vue-standalone";
    Preset["Nuxt"] = "nuxt";
    Preset["NuxtStandalone"] = "nuxt-standalone";
    Preset["NextJs"] = "next";
    Preset["NextJsStandalone"] = "nextjs-standalone";
    Preset["RemixMonorepo"] = "remix-monorepo";
    Preset["RemixStandalone"] = "remix-standalone";
    Preset["ReactNative"] = "react-native";
    Preset["Expo"] = "expo";
    Preset["Nest"] = "nest";
    Preset["Express"] = "express";
    Preset["React"] = "react";
    Preset["Vue"] = "vue";
    Preset["Angular"] = "angular";
    Preset["NodeStandalone"] = "node-standalone";
    Preset["NodeMonorepo"] = "node-monorepo";
    Preset["TsStandalone"] = "ts-standalone";
})(Preset || (exports.Preset = Preset = {}));
/**
 * This function is used to check if a preset is a known Nx preset.
 * @param preset
 * @returns true if the preset is a known Nx preset, false otherwise.
 */
function isKnownPreset(preset) {
    return Object.values(Preset).includes(preset);
}
exports.isKnownPreset = isKnownPreset;
